File: ecommerce-ts-express/services/orders/src/modules/order/order.service.ts
import * as grpc from '@grpc/grpc-js'
import {
  CreateOrderDto,
  Currency,
  OrderDto,
  OrderStatus,
  OrderWithProductDto,
  PaginatedResponse,
  ProductDto
} from '@repo/contracts'
import { getEnv } from '../../config/env.js'
import { AppDataSource } from '../../config/postgres.js'
import { ProductsGrpcClient } from '../../grpc/products.client.js'
import { OrderItemEntity } from '../order-item/order-item.entity.js'
import { OrderEntity } from './order.entity.js'
import { toOrderDto, toOrderWithProducts } from './order.mapper.js'
import { OrderRepo } from './order.repo.js'

export class OrderService {
  constructor(
    private readonly repo: OrderRepo,
    private readonly productsGrpcClient: ProductsGrpcClient
  ) {}

  async create(dto: CreateOrderDto): Promise<OrderDto> {
    if (dto.items.length === 0) throw new Error('items is required')

    return AppDataSource.transaction(async (manager) => {
      const orderRepo = manager.getRepository(OrderEntity)
      const itemRepo = manager.getRepository(OrderItemEntity)

      const currency = Currency.RUB // Упрощение

      const order = orderRepo.create({
        userId: dto.userId,
        status: OrderStatus.NEW,
        currency
      })

      const saved = await orderRepo.save(order)

      const items = dto.items.map((it) =>
        itemRepo.create({
          orderId: saved.id,
          productId: it.productId,
          quantity: it.quantity
        })
      )

      await itemRepo.save(items)

      const full = await orderRepo.findOne({
        where: { id: saved.id },
        relations: { items: true }
      })
      if (!full) throw new Error('Order not found after create')

      return toOrderDto(full)
    })
  }

  async getById(orderId: string): Promise<OrderDto> {
    const order = await this.repo.findById(orderId)
    if (!order) throw new Error('Order not found')
    return toOrderDto(order)
  }

  async listAll(params: {
    page: number
    limit: number
    offset: number
  }): Promise<PaginatedResponse<OrderDto>> {
    const { rows, total } = await this.repo.findAll({
      offset: params.offset,
      limit: params.limit
    })

    return {
      data: rows.map(toOrderDto),
      page: params.page,
      limit: params.limit,
      total
    }
  }

  async listAllHttp(params: {
    page: number
    limit: number
    offset: number
  }): Promise<PaginatedResponse<OrderWithProductDto>> {
    const { rows, total } = await this.repo.findAll({
      offset: params.offset,
      limit: params.limit
    })

    const ids = Array.from(
      new Set(rows.flatMap((o) => (o.items ?? []).map((i) => i.productId)))
    )

    const products = await this.getProductsBatchHttp(ids)

    return {
      data: rows.map((o) => toOrderWithProducts(o, products)),
      page: params.page,
      limit: params.limit,
      total
    }
  }

  private async getProductsBatchHttp(ids: string[]): Promise<ProductDto[]> {
    if (ids.length === 0) return []

    const env = getEnv()
    const base = env.productsBaseUrl

    const res = await fetch(`${base}/batch`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ ids })
    })

    if (!res.ok) {
      const text = await res.text()
      const err: any = new Error(
        `Products service error (${res.status}): ${text}`
      )
      err.statusCode = 502
      throw err
    }

    return (await res.json()) as ProductDto[]
  }

  async listAllGrpc(params: { page: number; limit: number; offset: number }) {
    const { rows, total } = await this.repo.findAll({
      offset: params.offset,
      limit: params.limit
    })

    const ids = Array.from(
      new Set(rows.flatMap((o) => (o.items ?? []).map((i) => i.productId)))
    )

    const products = await this.getProductsBatchGrpc(ids)

    return {
      data: rows.map((o) => toOrderWithProducts(o, products)),
      page: params.page,
      limit: params.limit,
      total
    }
  }

  private async getProductsBatchGrpc(ids: string[]): Promise<ProductDto[]> {
    if (ids.length === 0) return []

    try {
      return await this.productsGrpcClient.batch(ids, 2000)
    } catch (e: any) {
      const err: any = new Error(
        `Products gRPC error (${e?.code ?? 'unknown'}): ${e?.message ?? e}`
      )
      err.statusCode = 502

      // если хочешь различать timeout:
      if (e?.code === grpc.status.DEADLINE_EXCEEDED) {
        err.statusCode = 504
      }

      throw err
    }
  }
}

File: ecommerce-go-fiber/apps/orders/internal/service/order_service.go
package service

import (
	"context"
	"errors"
	"fmt"

	"github.com/ivanmonichev/ecommerce-go-fiber/apps/orders/internal/domain"
	"github.com/ivanmonichev/ecommerce-go-fiber/apps/orders/internal/mapper"
	"github.com/ivanmonichev/ecommerce-go-fiber/apps/orders/internal/repository"
	grpcclient "github.com/ivanmonichev/ecommerce-go-fiber/apps/orders/internal/transport/grpc_client"
	httpclient "github.com/ivanmonichev/ecommerce-go-fiber/apps/orders/internal/transport/http_client"
	"github.com/ivanmonichev/ecommerce-go-fiber/shared/contracts/common"
	"github.com/ivanmonichev/ecommerce-go-fiber/shared/contracts/orders"
	"github.com/ivanmonichev/ecommerce-go-fiber/shared/contracts/products"
)

var ErrItemsRequired = errors.New("items is required")
var ErrInvalidOrderStatus = errors.New("invalid order status")

type OrderService struct {
	repo               *repository.OrderRepository
	productsHTTPClient httpclient.ProductsHTTPClient
	productsGRPCClient grpcclient.ProductsGRPCClient
}

func NewOrderService(
	repo *repository.OrderRepository,
	productsHTTPClient httpclient.ProductsHTTPClient,
	productsGRPCClient grpcclient.ProductsGRPCClient,
) *OrderService {
	return &OrderService{
		repo:               repo,
		productsHTTPClient: productsHTTPClient,
		productsGRPCClient: productsGRPCClient,
	}
}

func (s *OrderService) Create(
	ctx context.Context,
	dto orders.CreateOrderDTO,
) (orders.OrderDTO, error) {
	if len(dto.Items) == 0 {
		return orders.OrderDTO{}, ErrItemsRequired
	}

	if dto.Currency == "" {
		dto.Currency = common.Currency("RUB")
	}

	created, err := s.repo.Create(ctx, dto)
	if err != nil {
		return orders.OrderDTO{}, err
	}

	return mapper.ToOrderDTO(created), nil
}

func (s *OrderService) GetByID(
	ctx context.Context,
	orderID string,
) (orders.OrderDTO, error) {
	order, err := s.repo.FindByID(ctx, orderID)
	if err != nil {
		return orders.OrderDTO{}, err
	}

	return mapper.ToOrderDTO(order), nil
}

func (s *OrderService) ListAllHTTP(
	ctx context.Context,
	query common.PaginationQuery,
) (common.PaginatedResponse[orders.OrderWithProductDTO], error) {
	query = query.Normalize()

	result, err := s.repo.FindAll(ctx, repository.PaginationParams{
		Offset: query.Offset(),
		Limit:  query.Limit,
	})
	if err != nil {
		return common.PaginatedResponse[orders.OrderWithProductDTO]{}, fmt.Errorf("find all orders: %w", err)
	}

	ids := collectProductIDs(result.Rows)

	productsList, err := s.productsHTTPClient.Batch(ctx, ids)
	if err != nil {
		return common.PaginatedResponse[orders.OrderWithProductDTO]{}, fmt.Errorf("fetch products by http: %w", err)
	}

	return s.buildOrdersWithProducts(query, result, productsList), nil
}

func (s *OrderService) ListAllGRPC(
	ctx context.Context,
	query common.PaginationQuery,
) (common.PaginatedResponse[orders.OrderWithProductDTO], error) {
	query = query.Normalize()

	result, err := s.repo.FindAll(ctx, repository.PaginationParams{
		Offset: query.Offset(),
		Limit:  query.Limit,
	})
	if err != nil {
		return common.PaginatedResponse[orders.OrderWithProductDTO]{}, fmt.Errorf("find all orders: %w", err)
	}

	ids := collectProductIDs(result.Rows)

	productsList, err := s.productsGRPCClient.Batch(ctx, ids)
	if err != nil {
		return common.PaginatedResponse[orders.OrderWithProductDTO]{}, fmt.Errorf("fetch products by grpc: %w", err)
	}

	return s.buildOrdersWithProducts(query, result, productsList), nil
}

/*
func (s *OrderService) ListAllGRPC(

	ctx context.Context,
	query common.PaginationQuery,

	) (common.PaginatedResponse[orders.OrderWithProductDTO], error) {
		query = query.Normalize()

		result, err := s.repo.FindAll(ctx, repository.PaginationParams{
			Offset: query.Offset(),
			Limit:  query.Limit,
		})
		if err != nil {
			return common.PaginatedResponse[orders.OrderWithProductDTO]{}, err
		}

		ids := collectProductIDs(result.Rows)

		productsList, err := s.productsGRPCClient.Batch(ctx, ids)
		if err != nil {
			return common.PaginatedResponse[orders.OrderWithProductDTO]{}, err
		}

		return s.buildOrdersWithProducts(query, result, productsList), nil
	}
*/
func (s *OrderService) buildOrdersWithProducts(
	query common.PaginationQuery,
	result repository.FindAllResult,
	products []products.ProductDTO,
) common.PaginatedResponse[orders.OrderWithProductDTO] {
	items := make([]orders.OrderWithProductDTO, 0, len(result.Rows))

	for _, row := range result.Rows {
		items = append(items, mapper.ToOrderWithProductsDTO(row, products))
	}

	return common.PaginatedResponse[orders.OrderWithProductDTO]{
		Data:  items,
		Total: result.Total,
		Page:  query.Page,
		Limit: query.Limit,
	}
}

func (s *OrderService) UpdateStatus(
	ctx context.Context,
	dto orders.OrderStatusUpdate,
) error {
	if !isValidStatus(dto.Status) {
		return ErrInvalidOrderStatus
	}

	return s.repo.UpdateStatus(ctx, dto.OrderID, dto.Status)
}

func isValidStatus(status orders.OrderStatus) bool {
	switch status {
	case orders.OrderStatusNew, orders.OrderStatusCanceled, orders.OrderStatusCompleted:
		return true
	default:
		return false
	}
}

func collectProductIDs(orders []domain.Order) []string {
	uniq := make(map[string]struct{})

	for _, order := range orders {
		for _, item := range order.Items {
			uniq[item.ProductID] = struct{}{}
		}
	}

	result := make([]string, 0, len(uniq))
	for id := range uniq {
		result = append(result, id)
	}

	return result
}
